### PR TITLE
Use single kms-plugin mock in unit and integration tests.

### PR DESCRIPTION
**Pull request by immutableT**
**What type of PR is this?**
/kind cleanup
**What this PR does / why we need it**:
Allow both integration and unit tests to leverage common implementation of a mock of kms-plugin.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 1]
```diff
@@ -69,6 +69,7 @@ func NewGRPCService(endpoint string, callTimeout time.Duration) (Service, error)
 			if err != nil {
 				klog.Errorf("failed to create connection to unix socket: %s, error: %v", addr, err)
 			}
+			klog.V(3).Infof("Successfully dialed Unix socket %v", addr)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: NewGRPCService logs at v4, and the dialer can be called more often... is v4 more appropriate?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kms_plugin_mock.go [hunk 2]
```diff
@@ -16,39 +16,38 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package master
+package envelope
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kms_plugin_mock.go]: suggest separating this into a testing or mock or fake package to make it really clear this isn't a real implementation

**Comment by liggitt** [comment]
fails on non-linux machines:

```
--- FAIL: TestIntermittentConnectionLoss (3.01s)
    grpc_service_unix_test.go:207: Connected to KMSPlugin
    grpc_service_unix_test.go:210: KMS Plugin is stopping
    grpc_service_unix_test.go:219: Sending encrypt request
    grpc_service_unix_test.go:232: failed to start test KMS provider server, error: failed to listen on the unix socket, error: listen unix @d3cef647-7b20-47c1-9466-6e3892e30a9a.sock: bind: address already in use
panic: Fail in goroutine after TestIntermittentConnectionLoss has completed
```

see https://github.com/kubernetes/kubernetes/pull/78985/files#diff-f28771f7c5b4d3a3e3a2ddeba979ac53R422

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go [hunk 3]
```diff
@@ -356,91 +371,31 @@ func destroyService(service Service) {
...
-type fakeKMSPlugin struct {
-	apiVersion string
-	server     *grpc.Server
-	sockFile   string
-}
-
-func (s *fakeKMSPlugin) Stop() {
-	// Stop the server
-	s.server.Stop()
-	// If this isn't a Linux abstract namespace socket, or if we're on a non-linux platform, clean up the socket file
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go]: this breaks unit tests of this package on darwin

**Comment by pduche** [comment]
Unsubscribe. 
I am getting too many emails. 
~ Duché Edouard ~  ~ La prière est mon chemin, la foi est ma source, et l’amour est ma vie ~     

    On Monday, October 14, 2019, 01:59:05 PM EDT, Kubernetes Prow Robot <notifications@github.com> wrote:  
 
 
@immutableT: The following tests failed, say /retest to rerun them all:

| Test name | Commit | Details | Rerun command |
| pull-kubernetes-e2e-gce-device-plugin-gpu | 53a5649 | link | /test pull-kubernetes-e2e-gce-device-plugin-gpu |
| pull-kubernetes-bazel-test | 53a5649 | link | /test pull-kubernetes-bazel-test |
| pull-kubernetes-e2e-gce-100-performance | 53a5649 | link | /test pull-kubernetes-e2e-gce-100-performance |
| pull-kubernetes-bazel-build | 53a5649 | link | /test pull-kubernetes-bazel-build |
| pull-kubernetes-kubemark-e2e-gce-big | 53a5649 | link | /test pull-kubernetes-kubemark-e2e-gce-big |
| pull-kubernetes-e2e-gce | 53a5649 | link | /test pull-kubernetes-e2e-gce |


Full PR test history. Your PR dashboard. Please help us cut down on flakes by linking to an open issue when you hit one in your PR.
 
Instructions for interacting with me using PR comments are available here. If you have questions or suggestions related to my behavior, please file an issue against the kubernetes/test-infra repository. I understand the commands that are listed here.

**Comment by pduche** [comment]
Please unsubscribe me.
Thank you!
~ Duché Edouard ~  ~ La prière est mon chemin, la foi est ma source, et l’amour est ma vie ~     

    On Monday, October 14, 2019, 01:59:05 PM EDT, Kubernetes Prow Robot <notifications@github.com> wrote:  
 
 
@immutableT: The following tests failed, say /retest to rerun them all:

| Test name | Commit | Details | Rerun command |
| pull-kubernetes-e2e-gce-device-plugin-gpu | 53a5649 | link | /test pull-kubernetes-e2e-gce-device-plugin-gpu |
| pull-kubernetes-bazel-test | 53a5649 | link | /test pull-kubernetes-bazel-test |
| pull-kubernetes-e2e-gce-100-performance | 53a5649 | link | /test pull-kubernetes-e2e-gce-100-performance |
| pull-kubernetes-bazel-build | 53a5649 | link | /test pull-kubernetes-bazel-build |
| pull-kubernetes-kubemark-e2e-gce-big | 53a5649 | link | /test pull-kubernetes-kubemark-e2e-gce-big |
| pull-kubernetes-e2e-gce | 53a5649 | link | /test pull-kubernetes-e2e-gce |


Full PR test history. Your PR dashboard. Please help us cut down on flakes by linking to an open issue when you hit one in your PR.
 
Instructions for interacting with me using PR comments are available here. If you have questions or suggestions related to my behavior, please file an issue against the kubernetes/test-infra repository. I understand the commands that are listed here.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go (hunk 3 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go]: Yes, I totally forgot about this.
Add the this logic into the CleanUp method.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kms_plugin_mock.go (hunk 2 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kms_plugin_mock.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 1 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go [hunk 4]
```diff
@@ -20,44 +20,54 @@ limitations under the License.
...
-	"google.golang.org/grpc"
+	mock "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing"
 
 	"k8s.io/apimachinery/pkg/util/uuid"
-	kmsapi "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1"
 )
 
+type testSocket struct {
+	path string
+	name string
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go]: nit: can we call this `endpoint` to match the previous usage and the param name to NewGRPCService?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go [hunk 5]
```diff
@@ -20,44 +20,54 @@ limitations under the License.
...
+}
+
+// newTestSocket constructs a unique name for a Linux Abstract Socket to be used in a test.
+// This package uses Linux Domain Sockets to remove the need for clean-up of socket files.
+func newTestSocket() *testSocket {
+	p := fmt.Sprintf("@%s.sock", uuid.NewUUID())
+
+	return &testSocket{
+		path: p,
+		name: fmt.Sprintf(fmt.Sprintf("unix:///%s", p)),
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go]: `fmt.Sprintf(fmt.Sprintf(`?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/kms_plugin_mock.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	listener           net.Listener
+	mu                 *sync.Mutex
+	lastEncryptRequest *kmsapi.EncryptRequest
+	inFailedState      bool
+	ver                string
+	socketPath         string
+}
+
+// NewBase64Plugin is a constructor for Base64Plugin.
+func NewBase64Plugin(socketPath string) (*Base64Plugin, error) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/kms_plugin_mock.go]: Just noticed calling this actually starts the listener. I wouldn't have expected that to happen until Start() was called

File: test/integration/master/kms_transformation_test.go [hunk 7]
```diff
@@ -197,23 +198,27 @@ resources:
...
-	pluginMock1, err := newBase64Plugin("@kms-provider-1.sock")
+	pluginMock1, err := mock.NewBase64Plugin("@kms-provider-1.sock")
 	if err != nil {
 		t.Fatalf("failed to create mock of KMS Plugin #1: %v", err)
 	}
 
-	go pluginMock1.grpcServer.Serve(pluginMock1.listener)
-	defer pluginMock1.cleanUp()
-	kmsPluginMustBeUp(t, pluginMock1)
+	go pluginMock1.Start()
```
Comment (liggitt) [inline test/integration/master/kms_transformation_test.go]: Start() launches a goroutine internally, so we can drop the `go` bit here, right? (applies to all callers)

File: test/integration/master/kms_transformation_test.go [hunk 8]
```diff
@@ -197,23 +198,27 @@ resources:
...
+	if err := mock.WaitForBase64PluginToBeUp(pluginMock1); err != nil {
+		t.Fatalf("Failed to start plugin #1, err: %v", err)
+	}
+	defer pluginMock1.CleanUp()
```
Comment (liggitt) [inline test/integration/master/kms_transformation_test.go]: defer cleanup right after starting

**Comment by immutableT** [comment]
@liggitt PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go]: still outstanding

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go [hunk 9]
```diff
@@ -139,11 +151,12 @@ func TestTimeouts(t *testing.T) {
 				// Simulating delayed start of kms-plugin, kube-apiserver is up before the plugin, if requested by the testcase.
 				time.Sleep(tt.pluginDelay)
 
-				f, err := startFakeKMSProvider(kmsapiVersion, socketName)
+				f, err := mock.NewBase64Plugin(socketName.path)
 				if err != nil {
 					t.Fatalf("failed to start test KMS provider server, error: %v", err)
 				}
-				defer f.Stop()
+				f.Start()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go]: check error return from Start()?

**Comment by liggitt** [comment]
two small comments, then lgtm

go ahead and squash as well

**Comment by liggitt** [comment]
should we adjust the klog adapter to make all info-level logs from etcd a higher verbosity (e.g. V3 or V4?)

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 10]
```diff
@@ -54,7 +54,7 @@ type gRPCService struct {
 
 // NewGRPCService returns an envelope.Service which use gRPC to communicate the remote KMS provider.
 func NewGRPCService(endpoint string, callTimeout time.Duration) (Service, error) {
-	klog.V(4).Infof("Configure KMS provider with endpoint: %s", endpoint)
+	klog.V(4).Infof("Configure KMS provider with testSocket: %s", endpoint)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: bad find/replace of endpoint with testSocket in messages/comments?

Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: yes.
Fixed.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 11]
```diff
@@ -84,15 +86,15 @@ func NewGRPCService(endpoint string, callTimeout time.Duration) (Service, error)
 	}, nil
 }
 
-// Parse the endpoint to extract schema, host or path.
+// Parse the testSocket to extract schema, host or path.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: sorry, a few other comment/message find/replace instances to clean up

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 12]
```diff
@@ -84,15 +86,15 @@ func NewGRPCService(endpoint string, callTimeout time.Duration) (Service, error)
...
 func parseEndpoint(endpoint string) (string, error) {
 	if len(endpoint) == 0 {
-		return "", fmt.Errorf("remote KMS provider can't use empty string as endpoint")
+		return "", fmt.Errorf("remote KMS provider can't use empty string as testSocket")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: here

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 13]
```diff
@@ -84,15 +86,15 @@ func NewGRPCService(endpoint string, callTimeout time.Duration) (Service, error)
...
 	}
 
 	u, err := url.Parse(endpoint)
 	if err != nil {
-		return "", fmt.Errorf("invalid endpoint %q for remote KMS provider, error: %v", endpoint, err)
+		return "", fmt.Errorf("invalid testSocket %q for remote KMS provider, error: %v", endpoint, err)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: here

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 11 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: No, I am sorry. I should have checked the rest of the file.
Fixed.

**Pull request merged**